engine/src/ast/function_expr.rs

```diff
@@ -91,13 +91,28 @@ impl<'s> FunctionCallExpr<'s> {
         self.args.iter().any(|arg| arg.uses(field))
     }
 
-    pub fn execute(&self, ctx: &'s ExecutionContext<'s>) -> LhsValue<'_> {
+    pub fn execute(&self, ctx: &'s ExecutionContext<'s>) -> Option<LhsValue<'_>> {
         self.function.implementation.execute(
-            self.args.iter().map(|arg| arg.execute(ctx)).chain(
-                self.function.opt_params[self.args.len() - self.function.params.len()..]
-                    .iter()
-                    .map(|opt_arg| opt_arg.default_value.as_ref()),
-            ),
+            self.args[..self.function.params.len()]
+                .iter()
+                .map(|arg| arg.execute(ctx))
+                .chain(
+                    self.args[self.function.params.len()..self.args.len()]
+                        .iter()
+                        .enumerate()
+                        .map(|(idx, arg)| {
+                            Some(arg.execute(ctx).unwrap_or_else(|| {
+                                self.function.opt_params[idx].default_value.as_ref()
+                            }))
+                        }),
+                )
+                .chain(
+                    self.function.opt_params[self.args.len() - self.function.params.len()..]
+                        .iter()
+                        .map(|opt_arg| Some(opt_arg.default_value.as_ref())),
+                )
+                .take_while(Option::is_some)
```

Comment:
These changes seem wrong... You are essentially saying that `f(a, b, c)` where `b` is missing should evaluate `f(a)` instead. That's counter-intuitive and might lead to incorrect results.

Reply:
My view on this is that a nil mandatory argument in the list should be seen as a short count of argument. Having an iterator that yields None one or multiple times and then yield value again its also counter-intuitive.

Reply:
Agreed, neither is perfect.

I think what we should do here is just ignore the function call altogether when one of the values is missing, and then follow regular comparison rules on the result (`true` for `!=`, `false` for others).

This way, function call is always well-defined and guaranteed to receive as many arguments as declared.

Reply:
That's initially what i wanted to implement, but, as you noted, it's a bit tricky here. I did not find a good way to do it while at the same time keeping the iterator. On the other side, passing nil to function will allow to make decision later about how exactly each function should behave (like `len` for example).

Reply:
I meant that i could not find a way to execute each argument to check for nil value without local allocation in the closure. I tried to pre-allocate an array of `Option<LhsValue>` but then either the closure become `FnMut` or each `execute` function becomes `&mut self`

Reply:
No, I got it.

In that case I think you can make `FunctionArgs` iterator item to be a `Result` instead where error is a missing field, and let functions decide what to do - propagate an error upwards or fall back somehow. This way you'll allow iterator to keep working even after a missing field.

Reply:
I updated the PR with `Result<LhsValue, ()>` type as function argument. I find the API a bit more confusing now but it surely provide maximum flexibility to let functions decide how to handle those kind of missing argument value.

Reply:
Btw, should a specific `MissingValueError` error be introduced?

Reply:
> Btw, should a specific MissingValueError error be introduced?

Yes. I thought we already had a corresponding error definition somewhere...